pkg/experiment/metastore/raftnode/node_admin.go

```diff
@@ -0,0 +1,135 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if err := n.raft.AddVoter(raft.ServerID(request.ServerId), raft.ServerAddress(request.ServerId), 0, 0).Error(); err != nil {
+		level.Error(n.logger).Log("msg", "failed to add node, error from raft", "node", request.ServerId, "err", err)
+		return nil, WithRaftLeaderStatusDetails(err, n.raft)
+	}
+
+	level.Info(n.logger).Log("msg", "node added", "id", request.ServerId)
+	return &raftnodepb.NodeChangeResponse{}, nil
+}
+
+func (n *Node) DemoteLeader(request *raftnodepb.NodeChangeRequest) (*raftnodepb.NodeChangeResponse, error) {
+	level.Info(n.logger).Log("msg", "demoting node", "id", request.ServerId)
+
+	// Send a round of heartbeats to confirm we are the leader. If we are not, the request will be retried on the leader node.
+	if err := n.raft.VerifyLeader().Error(); err != nil {
+		level.Error(n.logger).Log("msg", "failed to demote node, we are not the leader", "id", request.ServerId)
+		return nil, WithRaftLeaderStatusDetails(err, n.raft)
+	}
+
+	// Verify that we are on the same term as the one in the request. Otherwise, the request could be operating on stale information.
+	err := n.verifyCurrentTerm(request.CurrentTerm)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure we are demoting the node from the request (we can only demote ourselves)
+	if n.config.ServerID != request.ServerId {
+		level.Error(n.logger).Log("msg", "failed to demote node, the target node is not the leader", "id", request.ServerId)
+		return nil, fmt.Errorf("the target node is not the leader")
+	}
+
+	if err := n.raft.LeadershipTransfer().Error(); err != nil {
+		level.Error(n.logger).Log("msg", "failed to demote node, error from raft", "node", request.ServerId, "err", err)
+		return nil, WithRaftLeaderStatusDetails(err, n.raft)
+	}
+
+	level.Info(n.logger).Log("msg", "node demoted", "id", request.ServerId)
+	return &raftnodepb.NodeChangeResponse{}, nil
+}
+
+func (n *Node) PromoteToLeader(request *raftnodepb.NodeChangeRequest) (*raftnodepb.NodeChangeResponse, error) {
+	level.Info(n.logger).Log("msg", "promoting node", "id", request.ServerId)
+
+	// Send a round of heartbeats to confirm we are the leader. If we are not, the request will be retried on the leader node.
+	if err := n.raft.VerifyLeader().Error(); err != nil {
+		level.Error(n.logger).Log("msg", "failed to promote node, we are not the leader", "id", request.ServerId)
+		return nil, WithRaftLeaderStatusDetails(err, n.raft)
+	}
+
+	// Verify that we are on the same term as the one in the request. Otherwise, the request could be operating on stale information.
+	err := n.verifyCurrentTerm(request.CurrentTerm)
+	if err != nil {
+		return nil, err
+	}
+
+	// make sure we are not promoting ourselves
+	if n.config.ServerID == request.ServerId {
+		level.Error(n.logger).Log("msg", "failed to promote node, we cannot promote ourselves", "node", request.ServerId)
+		return nil, status.Error(codes.InvalidArgument, "a node cannot promote itself")
+	}
```

Comment:
It's very unlikely (if possible) that we get to "a node cannot promote itself" if this is the case (probably, the same for other handlers)

1. The identity check should fail if the leader has changed since we've read the cluster state the last time (to show in UI).
2. If leader changes before the check, `verifyCurrentTerm` will fail as the term increases when the old leader steps down.
3. If leader changes after the check and before the change is committed to the raft log, the raft fails to perform the operation and we redirect to the new leader. For the new leader the situation looks like p.1 and it fails before the identity check.

I'm wondering if we should swap those.

Reply:
It will currently happen if a promote request is sent to the current leader (e.g., due to a bug in the UI).

I thought about swapping the identity and term checks, but the errors for concurrent operations become misleading. E.g., if we promote a node that was promoted already in a previous request, we'll get an error "a node promote ourselves" as opposed to the "term check failed" error. In this scenario, the second error is actually more clear (you operated on stale information). 

Of course, an option is to not fail at all in this case, but then as I argued earlier, this would hide information from the operator that something didn't quite go as they thought.